### ISSUE TITLE
Potential misclassification in several cases on a real patient file on read level

### ISSUE THREAD

**Issue by pfeiferd** [DerrickWood/kraken2#712]
Hi,

this concerns real patient data (pubished here with the consent of the patient). The gzipped-fastq file has 700MB. My analysis referred to the taxids 1293575 and 47466 only. (So it might well be that similar misclassifications exist (regarding that fastq-file) with respect to other taxids.)

These are the counts from kraken2 with respect to classes:
taxid;classifications
1293575;1
47466;3
All 4 of them are apparently wrong (as checked via blasts). You can find the corresponding reads and respective kraken2 output below.

These are the counts on the kmer-level:
taxid;kmers
1293575;8
47466;73
Apparently, they are all wrong too (as checked via lookups in the corresponding refseq genome files) but this may be expected as @DerrickWood said in another reply due to minimizers and the probabilistic nature of the kraken2's hashtable.

Note these were the only hits for the 2 taxids regarding the patient's file - and none of them are correct. I used kraken2-2.1.2 (on PowerMac - but same on Linux, default settings) with k2_standard_08gb_20221209 as the database.

It seems that (frequent) mismatches on the kmer level may lead to misclassifications on the read level (at least sometimes) which in turn might lead to false conclusions... (?)

Here are the reads and the respecitve output from Kraken 2:

1)
@A01245:144:HMV7FDSX3:1:2341:32633:18364/2
TCCTATATGTGAATGAAGCAAGCATAATTTTCCTGAATATATTCTTCAAACATCCATCTATGCTAAAAGTATCAAATATGTATCCAAATCTTTATTTAATTATTTTATTATTTAATGATTTATCTTATAAAAAAAAAACACTGAAGAAGAA
+
FF,FFFFFFFFFFFFFF,FFF:,FFFFFFFF:F:,FF,FFFF,::F:FFF,FFF,F,F,F,,,,FFFF,,F,,,F,FF:,:F,:,FF,:,F,,F,::F,:F,,F::F::F:,:FF,,,:,:,F,FFF,F:F,F:,,,FF:,,,,:,F:,FF
C	A01245:144:HMV7FDSX3:1:2341:32633:18364/2	1293575	151	0:1 9606:3 0:92 1293575:5 0:16

2)
@A01245:144:HMV7FDSX3:2:1332:10845:27602/2
GGGGGGGGGGGTTGGGTGGGTTTTGTTTGTTTTTTTTTTAAAAATAATATTTATTAAAAATAAAAAAATAAAAATAAAAAAAAAATAAAAAAAAAAAAATAATAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
+
F:FF,F,:,,:,,,,,,,,,,,,,,,,:,,,:,,:,,,,,,F,,,,:,,,,,,,F,,:,,,,:,,,,,,,F,,,,,,,,:,,:::,F,,,F,,,F:,:,,:,,F,F,,:F,:,FF,,,F,:::FFFF::F:F,FFFFF:FFFFFFFFFFFF
C	A01245:144:HMV7FDSX3:2:1332:10845:27602/2	47466	151	0:43 47466:5 0:14 9606:2 0:53

3)
@A00925:335:HMFMVDSX5:2:1277:5312:11036/2
CCACCACACCCGGCTGTTTTTTTTTTTTTTTTTTTATTTTTAATAAATAAGGGGTTTTTGCAAATTTGCGGTGGTGGTCAAGAAAAACTGAACTCAGGTTATACACCAGCCGCCGCCAACCAAATTGTTGGGGTTAAAAGTGTTTTCCA
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFF:F,,F,FFF,:FFF,F,,FFF,FFF,:::,,:F,:F,,,F,,,F,,,,,,F,:,F,,,,,,,,,:F,,:,,F:FFF,:,F,::F:,:F,FF,:,,F,:F,,FFF,,F,,F,,,:F,
C	A00925:335:HMFMVDSX5:2:1277:5312:11036/2	47466	149	0:7 9606:1 0:8 47466:3 0:96

4)
@A00925:335:HMFMVDSX5:4:1420:1759:34194/2
TCAATTATCTTTGCCGGCTATTATTGATTTAGTTTTTTATTCTTTTTATTTATTTTTTAGAAAGAGACTACCACTTTAAACCAACCGACATTCCAATATCACTAAAACCGCACAACCAAATAAACCCCAACCCGGATAACACCCAACTACA
+
FFF::FF,,:FFFF:F,:,,F::FF,,,,F,FF,,F:,FFF,F,,F:,:FF,FF,,,F,,F:,,,,,,,,,F,,,,,:,,F,,,F,,,:F,,,FF:,:F,,,FFF,,,,F:F:,,:F,,F,,F,F,F,FFF:F:F,,,:,:,,,FF,,,,F
C	A00925:335:HMFMVDSX5:4:1420:1759:34194/2	47466	151	0:17 47466:5 0:2 9606:3 0:90

Regards,
Daniel

**Comment by salzberg** [comment]
Yes, Kraken2 has a very small false positive rate, which can be a problem when looking for a very small number of microbial reads in a human sample (as we do when we are trying to diagnose infections). That's why I always use KrakenUniq for this application, not Kraken2. The DB is much larger, but with the new --preload-size option, that's no longer a problem.

**Comment by pfeiferd** [comment]
Thanks Steven. I wondering how "small" the fp rate is for practical concerns. Sounds like KrakenUnique is the "only way" for pathogen analysis? In other words: kraken2 (by itself) is not the right tool for it?

**Comment by salzberg** [comment]
> Thanks Steven. I wondering how "small" the fp rate is for practical concerns. Sounds like KrakenUnique is the "only way" for pathogen analysis? In other words: kraken2 (by itself) is not the right tool for it?

In my opinion, krakenuniq is the only way, yes - for this application. Usually the number of FPs from Kraken2 for any species is tiny, typically 1-5 reads (as you seem to have found). So for most analyses, reads at that level are ignored anyway and it doesn't matter. But if you are looking for just a few reads out of millions, as we sometimes have done in our infectious disease diagnosis work, then those FPs are distracting, because we have to check them all out. KrakenUniq never does this - it's deterministic, so if you see a match, then you can be 100% certain that the read must share a 31-mer with the genome it matched.

**Comment by pfeiferd** [comment]
Steven, I appreciate you openness. The krakenuniq paper mentions reasons for error in conjunction with kraken2 such as sample contamination and "low-complexity regions". I think the inherent FP rate of kraken2 could have been highlighted more explicitely as a source of error (or have I overread it?). Afterall, a whole new industry is gonna build on these kind of tools and startups or patients should not have to learn this "the hard way". (I am not part of that scene though.)
May I say that even a sort disclaimer in the manual could be useful.

**Comment by salzberg** [comment]
Well, initially the Kraken2 team (I wasn't on that paper) didn't realize that the FP rate was high enough to matter, because they weren't looking at using it for infectious disease diagnosis. I found out this was happening in my own tests of Kraken2 and told them, and they added some parameters that reduce the FP rate significantly, but still not to zero. (By FP here, I mean a report of a hit when there was no actual k-mer match at all.) Kraken1 and KrakenUniq simply never do this. In our recent very short bioRxiv preprint on the low-memory option for KrakenUniq (published in JOSS), we mentioned this problem with Kraken2: I'll paste the text below.
Quote from [https://www.biorxiv.org/content/10.1101/2022.06.01.494344v1]: "The newer Kraken2 system (Wood et al., 2019) achieves a significantly lower memory footprint by using probabilistic data structures to reduce the database size, at the cost of slightly lower accuracy than KrakenUniq. This reduction in accuracy includes a very small but non-zero false positive rate (i.e., where the system incorrectly reports that a k-mer is present in a particular genome), which is problematic for certain applications that require very high precision. In particular, when metagenomic sequencing is used for the diagnosis of infections in a clinical setting (Salzberg et al., 2016), the pathogen of interest might be detected from just a handful of reads. In that scenario, even a few false positives can be confusing, and KrakenUniq is the preferred method rather than Kraken2."

**Comment by josruirod** [comment]
Thank you for the useful discussion. Related to this, I just want to ask wheter the considerations above on the FP affect all versions and modes of kraken. If this is relevant, as for diagnosis, krakenuniq is 100% the way to go? Or the flag --report-minimizer-data does the same according to the manual?

Thanks

**Comment by salzberg** [comment]
> Thank you for the useful discussion. Related to this, I just want to ask wheter the considerations above on the FP affect all versions and modes of kraken. If this is relevant, as for diagnosis, krakenuniq is 100% the way to go? Or the flag --report-minimizer-data does the same according to the manual?
> 
> Thanks

No, that is oversimplifying things far too much. Kraken2 is great and should be the way to go for most applications. However, for diagnosing infections from human samples - a very specific clinical application that I've been working on for the past 7 years - krakenuniq is better, because you know with 100% certainty that every match means the k-mer is present in the genome that it reports. You still get false positives from low-complexity hits, that doesn't change. And Kraken2 uses far less RAM for its database, which is a big advantage for most metagenomics applications.

**Comment by pfeiferd** [comment]
Very well spoken Steven - thaaanks - especially for the quotation!
